text/0000-generalized-arity-tuples.md

```diff
@@ -0,0 +1,277 @@
+
+## Why are we already done here?
+
+Consider the type `(Option<i32>, (Option<bool>, Option<&str>))` from the requirements above. This type is just syntactic sugar for the type `Tuple<Option<i32>, Tuple<Tuple<Option<bool>, Tuple<Option<&str>, ()>>, ()>>`.
+
+Now, apply the provided trait implementations step by step on the desugared type. The resulting associated type `Join::Joined` evaluates to `Tuple<i32, Tuple<Tuple<bool, Tuple<&str, ()>>, ()>>`. But this is just the desugared version of `(i32, (bool, &str))`.
+
+This illustrates how the mechanims works on the type level.
+
+## More examples / Advanced type mappings
+
+Generalized implementations on `Tuple<HEAD, TAIL>` are not restricted to mappings between tuples of the same shape. The following example demonstrates how a `last` function can be realized that works with any tuple:
+
+```rust
+trait Last {
+    type Last;
+
+    fn last(self) -> Self::Last;
+}
+
+impl<ELEM> Last for (ELEM,) {
+    type Last = ELEM;
+
+    fn last(self) -> Self::Last {
+        self.elem
+    }
+}
+
+impl<ELEM, TAIL: Last> Last for Tuple<ELEM, TAIL> {
+    type Last = TAIL::Last;
+
+    fn last(self) -> Self::Last {
+        self.tail.last()
+    }
+}
+```
+
+Correcty, the compiler rejects empty tuples while tuples of other sizes are accepted:
+
+```rust
+().last();       // does not compile: no method named `last` found for type `()` in the current scope
+(1,).last();     // returns 1
+(1, "two").last(); // returns "two"
+etc.
+```
+
+The last example demonstrates how every second element of a tuple can be removed:
+
+```rust
+trait Halve {
+    type Output;
+
+    fn halve(self) -> Self::Output;
+}
+
+impl Halve for () {
+    type Output = ();
+
+    fn halve(self) {}
+}
+
+impl<ELEM1, ELEM2, TAIL: Halve> Halve for Tuple<ELEM1, Tuple<ELEM2, TAIL>> {
+    type Output = Tuple<ELEM1, TAIL>;
+
+    fn halve(self) -> Self::Output {
+        Tuple::new(self.elem, self.tail.tail)
+    }
+}
+```
+
+Results:
+
+```rust
+().halve();                   // returns ()
+(1,).halve()                  // does not compile: no method named `halve` found for type `(i32,)` in the current scope
+(1, "two").halve();           // returns (1,)
+(1, "two", 3.0, '4').halve(); // returns (1, 3.0)
+etc.
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The idea of this RFC to leave the tuple notation as is but to treat tuple type expressions as aliases according to the following pattern:
+
+```rust
+type () = (); // Not really an alias. Written down for completeness.
+type (A,) = Tuple<A, ()>;
+type (A, B) = Tuple<A, Tuple<B, ()>>;
+type (A, B, C) = Tuple<A, Tuple<B, Tuple<C, ()>>>;
+type (A, (B, C)) = Tuple<A, Tuple<Tuple<B, Tuple<C, ()>>, ()>>
+etc.
+```
+
+where `Tuple` is a new struct located in `std::ops` with the following definition:
+
+```rust
+struct Tuple<ELEM, TAIL> {
```

Comment:
So this is essentially a HList: https://docs.rs/frunk/0.3.0/frunk/hlist/index.html.

- The good aspect of this is that you are essentially removing [`ty::TyKind::Tuple`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/enum.TyKind.html#variant.Tuple). The rules about unsized types in the last element should just fall out from structs. Overall, this is a substantial reduction in the amount of structural typing in the type system, which is a good thing. Instead, `Tuple<H, T>` is nominally typed and `(T0, ..., Tn)` is just sugar. You may want to extend the rationale with a note about the benefits of this simplification.

   - Also please make a note of `Tuple` becoming a `#[lang = "tuple"]` item by attaching the attribute here.

- On the other hand, this also means that the compiler is no longer free to apply layout optimizations where fields are reordered. E.g. today, the compiler is free to lay `(A, B, C, D)` out as laid out as `A D C B`. After introducing `struct Tuple<H, T>`, the compiler can no longer do that because it is now possible to take a reference to `tup.tail`.
   - See https://github.com/rust-lang/unsafe-code-guidelines/issues/105 for a discussion.
   - As a consequence, this may inflate the memory footprint of using tuples and other perf characteristics as well.
   - This problem could however be solved by not exposing `pub tail` like this and not exposing a method to take a reference to `tail` either. Moreover, we might want a sort of attribute `#[non_referenceable]` on fields which would prevent you from taking a reference `&tup.tail`. This would allow the compiler to reorder as it sees fit. This would be beneficial in other circumstances where refactoring a data type into pieces is currently not zero cost, an unfortunate situation.
    - However, by not making the tail referenceable, you give up on certain transformations that `frunk` provides.

Reply:
In a discussion with @oli-obk, they noted that `#[non_referenceable] pub tail: Tail` poses a problem wrt. to implementing `PartialEq` and similar traits when you want to do it recursively and generally for all tuples.

Reply:
More details:

```rust
impl PartialEq for () {
    fn eq(&self, other: &Self) -> bool {
        true
    }
}
impl<ELEM: PartialEq, TAIL: PartialEq> PartialEq for Tuple<ELEM, TAIL> {
    fn eq(&self, other: &Self) -> bool {
        self.elem == other.elem && self.tail == other.tail
    }
}
```

The `self.tail == other.tail` is essentially `PartialEq::eq(&self.tail, &other.tail)`, which would violate `#[non_referenceable]`.

Reply:
If I get this right, the compiler still has the flexibility to lay out `Tuple<ELEM, TAIL>` as either `ELEM TAIL` or `TAIL ELEM`. So `(A, B, C, D)` could become `A B C D` or `B C D A` or `A C D B` but not `B A C D`.

But, indeed, this is a hard restriction which might increase the memory footprint of every tuple.

This problem could be mitigated if the tuple representation was changed to a tree structure, e.g. `Tuple<ELEM, LEFT, RIGHT>`. In this way, the compiler could regain some control about the memory layout. In return, the compiler would need to match `Tuple<Elem1, Elem2, ()>` with `Tuple<Elem1, (), Elem2>` or wouldn't it? My first feeling is that this solution is bad just because it is not simple enough.

Reply:
Maybe I misunderstood something in previous discussions, but it seemed like we already knew there's a fundamental choice we have to make here between:

- the type lists used by variadic generics are identical to tuple types, and can compile away to nothing because tuples "don't exist at runtime", which rules out things like sub-tuples having addresses
- the type lists used by variadic generics are distinct from tuple types, so there is a certain amount of unfortunate duplication going on, but we get to make guarantees about tuple layout/addressability/etc

And any future variadic generics / generic tuples proposal would simply have to pick one of these and argue convincingly for it, but making up our minds on this was the main thing blocking progress on variadics. Is that no longer the case?

Reply:
Well there's a third option:
+ Go ahead with this proposal either with `#[non_referenceable]` or without, and then never add "variadic generics".

Reply:
I in my understanding, solving the variadic tuple problem is equivalent to solving the variadic generic problem. I would even go so far as to say that you do not need variadic generics if you have variadic tuples.

@Ixrec am not aware that the variadic generic problem has a final solution yet, has it?

Reply:
Yeah, it's definitely not solved yet. The whole point of my last comment was that any proposal for variadic tuples effectively is a proposal to solve the variadic generics problem. In other words, even if the proposed solution is simply that variadic tuples are enough, this needs to be made explicit, and then it has to be argued that "full" variadic generics are unnecessary or not worth it (afaik no one's made that argument before; maybe I could be convinced).

Reply:
I _believe_ that variadic generics might not be strictly necessary since you could model every type `Generic<A, ...>` with variadic generics simply as `Generic<T> where T: TraitThatIsValidForAllTuples`. The transition to a real generic notation could be done using a syntactic sugar approach.